data/data-pipeline/data_pipeline/content/config/generate_data_dictionary.py

```diff
@@ -0,0 +1,185 @@
+    "All percentiles are floored (rounded down to the nearest percentile). "
+    + "For example, 89.7th percentile is rounded down to 89 for this field."
+)
+LOW_PERCENTILE_EXPLANATION = "This percentile is reversed, meaning the lowest raw numbers become the highest percentiles."
+ISLAND_AREAS_EXPLANATION = (
+    "Because not all data is available for the Nation, Puerto Rico, "
+    + "and the Island Territories, this uses different underlying data for the islands."
+)
+
+
+OUTPUT_PATH = (
+    settings.APP_ROOT / "data" / "score" / "downloadable" / "codebook.csv"
+)
+
+
+def _set_up_infrastructure(
+    yaml_path: pathlib.Path,
+    fields_to_store: list,
+):
+    """helper function to parse yaml"""
+    yaml_contents = yaml.safe_load(
+        open(yaml_path, "r", encoding="utf-8").read()
+    )
+    if "sheets" in yaml_contents:
+        fields = []
+        for sheet in yaml_contents["sheets"]:
+            # excel yaml has ability for many sheets
+            fields.extend([single_field for single_field in sheet["fields"]])
+    else:
+        # other files do not
+        fields = yaml_contents["fields"]
+
+    to_frame_dict = {CEJST_SCORE_COLUMN_NAME: []} | {
+        field: [] for field, _ in fields_to_store
+    }
+
+    return to_frame_dict, fields
+
+
+def _create_df_from_yaml(
+    yaml_path: pathlib.Path,
+    fields_to_store: list,
+    native_score_column: str = CEJST_SCORE_COLUMN_NAME,
+):
+    """helper function to create a dataframe from yaml"""
+    to_frame_dict, fields = _set_up_infrastructure(
+        yaml_path=yaml_path,
+        fields_to_store=fields_to_store,
+    )
+    for details in fields:
+        to_frame_dict[CEJST_SCORE_COLUMN_NAME].append(
+            details[native_score_column]
+        )
+        for column_name, dict_field in fields_to_store:
+            try:
+                to_frame_dict[column_name].append(details[dict_field])
+            except KeyError:
+                assert column_name != CEJST_SCORE_COLUMN_NAME
+                to_frame_dict[column_name].append(np.nan)
+    return pd.DataFrame(to_frame_dict).set_index(CEJST_SCORE_COLUMN_NAME)
+
+
+def _get_datatype(
+    input_name: str,
+    input_type: str,
+    percentile_string: str = PERCENTILE_STRING,
+    loss_rate_string: str = LOSS_RATE_STRING,
+):
+    """helper to convert datatype"""
+    if percentile_string in input_name:
+        return "percentile"
+    elif loss_rate_string in input_name:
+        return "rate"
+    else:
+        return input_type
+
+
+def _get_calculation_notes(column_name):
+    """produces calculation notes"""
+    calculation_notes = []
+    if PERCENTILE_STRING in column_name:
+        calculation_notes += [PERCENTILE_EXPLANATION]
+    if LOW_STRING in column_name:
+        calculation_notes += [LOW_PERCENTILE_EXPLANATION]
+    if ISLAND_STRING in column_name:
+        calculation_notes += [ISLAND_AREAS_EXPLANATION]
+    return " ".join(calculation_notes)
+
+
+if __name__ == "__main__":
```

Comment:
I wonder if we could wrap all of this overall logic into a single function, and then calling that function part of `generate_score_post` in `application.py` so that this file gets automatically generated along with the rest of the downloads whenever we run those steps as part of existing processes? Would be easier for others to use than remembering to one-off run this script.

Reply:
It can also be helpful to put in some nice logger statements so that when someone is running this from the command line, they can "follow along". 

E.g., ``logger.info(f"Loading CSV configuration from file path `{CSV_CONFIG}`.")``. Same for Excel and Score. And then ``logger.info(f"Writing data dictionary to file path `{OUTPUT_PATH}`.")``. These loggers are helpful in debugging later and understanding what was done when.

Reply:
this makes a lot of sense to me! The code is pretty quick on my machine, so I didn't think to add logging but am happy to do so. Similarly, I like the idea of adding it to `generate_score_post`!

Reply:
One question I have now that this is in `generate_score_post` is if we need tests for it? When it was a one-off thing, I was less concerned about it breaking. Here, I'd envision the test being a snapshot, but a snapshot of the whole codebook. @lucasmbrown-usds wdyt?

Reply:
I always think we need tests, obviously! Would those be easy to add? Or you can create a backlog ticket for it, but as mentioned on the other PR, tickets for tests alone can be a real drag to add at a later date. 

Snapshot of the whole codebook makes sense to me! And would be v easy to update with our snapshot code.

As for logging, regardless of how long it takes to run they can be helpful on a remote server for pinpointing how far a process got before it crashed, and can be helpful to someone reading that output to know exactly what is generally happening under the hood during a process even when it's successful.

Reply:
I added an issue to the product backlog for now; I'll aim to take care of it sooner rather than later.